### DISCUSSION TITLE
Improved Webhook delegation

### DISCUSSION THREAD

**Discussion by doutatsu** [pay-rails/pay#686] [General] [answered]
I've been working through building logic for hooking into Pay webhooks, but I find it slightly awkward. Perhaps I missed something in the documentation, but I have two issues:

First of all, subscribing to a lot of events is pretty awkward:

```rb
Pay::Webhooks.configure do |events|
  events.subscribe "stripe.customer.subscription.created", Billing::HandleWebhookService.new
  events.subscribe "stripe.customer.subscription.updated", Billing::HandleWebhookService.new
  events.subscribe "stripe.customer.subscription.deleted", Billing::HandleWebhookService.new
  events.subscribe "stripe.payment_method.attached", Billing::HandleWebhookService.new
  events.subscribe "stripe.payment_method.detached", Billing::HandleWebhookService.new
  events.subscribe "stripe.payment_method.updated", Billing::HandleWebhookService.new
end
```

Is there no way to just provide a list and subscribe to a single class for handling a subset of webhooks? The only improvement I could do for now, is to iterate as such:

```rb
Pay::Webhooks.configure do |events|
  PAY_WEBHOOKS.each do |event_name|
    events.subscribe event_name, Billing::HandleWebhookService.new  
  end
end
```

Second issue I am having, is working with different classes that don't implement `call` method. For example, I want to call a Sidekiq worker to do some logic, but I can't use it, as it requires calling `perform_async` method. As a result, I have to make this awkward intermediary service to make it work:

```rb
Pay::Webhooks.configure do |events|
  PAY_WEBHOOKS.each do |event_name|
    events.subscribe event_name, Billing::HandleWebhookService.new  
  end
end

module Billing
  class HandleWebhookService
    def call(event)
      Webhooks::Billing::HandleUpdatesWorker.perform_async(event.to_json)
    end
  end
end
```

It would have been nicer to just directly call the worker or any class in general, that doesn't implement `call`:

```rb
Pay::Webhooks.configure do |events|
  PAY_WEBHOOKS.each do |event_name|
    events.subscribe event_name, SomeWorker.perform_async
    events.subscribe event_name, Class.start_sync  
  end
end
```
---
Maybe all these things are possible already, but I couldn't find anything obvious in the documentation or looking through the source code

**Comment by excid3** [comment, marked as answer]
1. That's just how ActiveSupport::Notification works. You subscribe to each event that you want to listen to and tell it where to send the event.
2. Just pass in a block.

See the ActiveSupport::Notification documentation for other examples of what you can do with it. 👍

**Reply by doutatsu** [reply]
wow rapid response @excid3, thanks for replying. I have looked at the documentation and even tried doing something like this:
```rb
Pay::Webhooks.configure do |events|
  PAY_WEBHOOKS.each do |event_name|
    events.subscribe event_name do |name, start, finish, id, payload|
      Webhooks::Billing::HandleUpdatesWorker.perform_async(payload.to_json)
    end
  end
end
```

Assuming the interface is the same, but that didn't work 🤔 What I am doing wrong here?

**Reply by excid3** [reply]
What didn't work exactly?

**Reply by doutatsu** [reply]
aha I gave up too soon... Based on ActiveSupport::Notification documentation, I presumed you had to pass through all `|name, start, finish, id, payload|` properties, which resulted in `null` being passed to the worker. Reason being there is only one argument in the Pay::Webhook itself (which should have been obvious, based on the behaviour I observed so far!), which I've now realised and fixed

So this configuration works:
```rb
Pay::Webhooks.configure do |events|
  PAY_WEBHOOKS.each do |event_name|
    events.subscribe event_name  do |event|
      Webhooks::Billing::HandleUpdatesWorker.perform_async(event.to_json)
    end
  end
end
```

Sorry about that 😅

**Reply by excid3** [reply]
Ah yeah, the Rails examples pass more information. We just pass the event.
